### PR TITLE
Use new PDF generator service for generating PDF receipts

### PR DESCRIPTION
**Pull request by RonnyB71**
## Description
This feature uses the new PDF Generator microservice when generating receipt PDF's. The new service receives a url to the application and generates the PDF based on the content of that url.

The page rendering the html to form the pdf is automatically created based on the layout(s) used with a print friendly css and respecting the excludeFromPdf setting on components as well as any custom implementations of IPdfFormatter. 

## Related Issue(s)
- #153 

## Verification
- [x] **Your** code builds clean without any errors or warnings
- [x] Manual testing done (required)
- [x] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [x] All tests run green

## Documentation
- [ ] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

### CODE REVIEW

File: src/Altinn.App.Core/Internal/Pdf/IPdfService.cs [hunk 1]
```diff
@@ -7,13 +11,18 @@ namespace Altinn.App.Core.Internal.Pdf
     /// </summary>
     public interface IPdfService
     {
-        /// <summary>
-        /// Generates the PDF based on the current data and stores it
```
Comment (tjololo) [inline src/Altinn.App.Core/Internal/Pdf/IPdfService.cs]: Summary removed by accident?

File: src/Altinn.App.Core/Internal/Pdf/IPdfGeneratorClient.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+/// Defines the required operations on a client of the PDF generator service.
+/// </summary>
+public interface IPdfGeneratorClient
+{
+    /// <summary>
+    /// Generates a PDF.
+    /// </summary>
+    /// <returns>A stream with the binary content of the generated PDF</returns>
+    Task<Stream> GeneratePdf(Uri uri, CancellationToken ct);
+}
```
Comment (tjololo) [inline src/Altinn.App.Core/Internal/Pdf/IPdfGeneratorClient.cs]:
```suggestion
}

```

File: src/Altinn.App.Core/Infrastructure/Clients/Pdf/PdfGeneratorClient.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            ex.Data.Add("responseContent", content);
+            ex.Data.Add("responseStatusCode", httpResponseMessage.StatusCode.ToString());
+            ex.Data.Add("responseReasonPhrase", httpResponseMessage.ReasonPhrase);
+
+            throw ex;
+        }
+
+        return await httpResponseMessage.Content.ReadAsStreamAsync(ct);
+    }
+}
```
Comment (tjololo) [inline src/Altinn.App.Core/Infrastructure/Clients/Pdf/PdfGeneratorClient.cs]:
```suggestion
}

```

File: src/Altinn.App.Core/Internal/Pdf/PdfGeneratorSettings.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    /// The name of a DOM element to wait for before triggering PDF-generator.
+    /// </summary>
+    public string WaitForSelector { get; set; } = "#readyForPrint";
+
+    /// <summary>
+    /// The number of milliseconds the PDF-generator should wait for a page to render. Default is 5000.
+    /// </summary>
+    /// <remarks>This will be ignored if <see cref="WaitForSelector"/> has been assigned a value.</remarks>
+    public int WaitForTime { get; set; } = 5000;
+}
```
Comment (tjololo) [inline src/Altinn.App.Core/Internal/Pdf/PdfGeneratorSettings.cs]:
```suggestion
}

```

File: test/Altinn.App.Core.Tests/Mocks/RequestInterceptor.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        }
+
+        internal class RequestInterceptorException : Exception
+        {
+            internal RequestInterceptorException(string? message) : base(message)
+            {
+            }
+        }
+    }
+}
```
Comment (tjololo) [inline test/Altinn.App.Core.Tests/Mocks/RequestInterceptor.cs]:
```suggestion
}

```

File: test/Altinn.App.Core.Tests/Internal/Pdf/PdfServiceTests.cs [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+                s => s.InsertBinaryData(
+                    It.Is<string>(s => s == instance.Id),
+                    It.Is<string>(s => s == "ref-data-as-pdf"),
+                    It.Is<string>(s => s == "application/pdf"),
+                    It.Is<string>(s => s == "not-really-an-app.pdf"),
+                    It.IsAny<Stream>()),
+                Times.Once);
+        }
+    }
+}
```
Comment (tjololo) [inline test/Altinn.App.Core.Tests/Internal/Pdf/PdfServiceTests.cs]:
```suggestion
}

```

**Review by tjololo — APPROVED** [review]
Can't find any more nitpick :godmode:

File: src/Altinn.App.Core/Internal/Pdf/IPdfService.cs (hunk 1 above)
Comment (RonnyB71) [reply src/Altinn.App.Core/Internal/Pdf/IPdfService.cs]: Yepp 🙈

**Comment by RonnyB71** [comment]
> Can't find any more nitpick :godmode:

Getting spoiled when you also includes fixes for the nitpicking :-D

**Pull request merged**